lib/WUI/wui.cpp

```diff
@@ -0,0 +1,620 @@
+            if (now - last_poll >= LOOP_EVT_TIMEOUT) {
+                last_poll = now;
+                events |= EspData | EthData | TriggerNtp;
+            }
+
+            if (events & CoreInitDone) {
+                // Chain the first reconfiguration to bring everything up as needed.
+                // (No need to go through the notification.
+                events |= Reconfigure;
+                initialized = true;
+                // TODO: Publish to the world we are initialized. Maybe something
+                // (connect) wants to wait for that.
+            }
+
+            // Note: This is allowed even before we are fully initialized. This
+            // is on purpose, because the initialization of ESP absolutely
+            // needs the communication.
+            //
+            // That's OK, because internally it can protect itself, it tracks
+            // if it is or isn't initialized.
+            if (events & EspData) {
+                events &= ~EspData;
+                espif_input_once(&ifaces[NETDEV_ESP_ID].dev);
+            }
+
+            if (!initialized) {
+                // Keep the events unprocessed for now, until the init is fully
+                // done.
+                continue;
+            }
+
+            if (events & Reconfigure) {
+                // Invoked directly after start and every time something in eeprom changes.
+                reconfigure();
+            }
+
+            if (events & EthInitDone) {
+                post_init(ifaces[NETDEV_ETH_ID]);
+            }
+
+            if (events & EspInitDone) {
+                post_init(ifaces[NETDEV_ESP_ID]);
+            }
+
+            if (events & EthData) {
+                ethernetif_input_once(&ifaces[NETDEV_ETH_ID].dev);
+            }
+
+            if (events & TriggerNtp) {
+                // TODO: This does some code gymnastics inside to track changes
+                // of network configuration. Consider cleaning that up and
+                // integrating into some kind of up/down mechanism.
+                sntp_client_step();
+            }
+
+            events = 0;
+        }
+    }
+    static void task_main(const void *) {
+        // Initialize our own thread ID.
+        NetworkState state;
+        state.run();
+    }
+
+    template <class F>
+    static void with_iface(uint32_t netdev_id, F &&f) {
+        // TODO: Lock!
+        NetworkState *state = instance;
+        if (state != nullptr && netdev_id < state->ifaces.size()) {
+            f(state->ifaces[netdev_id].dev, *state);
+        }
+    }
+    bool netif_link(uint32_t netdev_id) {
+        if (netdev_id == NETDEV_ETH_ID) {
+            return ethernetif_link(&ifaces[NETDEV_ETH_ID]);
+        }
+        if (netdev_id == NETDEV_ESP_ID) {
+            return espif_link();
+        }
+        assert(0);
+        return false;
+    }
+
+public:
+    NetworkState() {
+        network_task = osThreadGetId();
+        assert(instance == nullptr);
```

Comment:
Wouldn't have helped much. 

That whole singleton-ish approach is needed to signal the right _thread_ about interrupts and stuff. That is, there needs to be the thread ID somewhere (which is now inside the instance and that one doesn't need to be atomic because it's only available through that „outer“ instance atomic and ready at the point of publication ‒ but if changed, the need for atomic variable and for chechking „is it set already?“ would just move into the thread ID instead. And that would lead to the fact I need to bunch of static methods outside and they would no longer have access to private fields…

I feel like there's no _elegant_ way to go around signalling the right (single) thread from interrupts while the interrupts may happen even before the thread is started and ready. Anyway, if you believe the Mayer's singleton way would result in more readable code (considering the above problem), I can give it a try.